src/sentry/utils/performance_issues/performance_detection.py

```diff
@@ -1395,6 +1403,202 @@ def _is_file_io_on_main_thread(self, span: Span) -> bool:
         return data.get("blocked_ui_thread", False) is True
 
 
+class MNPlusOneState(ABC):
+    """Abstract base class for the MNPlusOneDBSpanDetector state machine."""
+
+    @abstractmethod
+    def next(self, span: Span) -> Tuple[MNPlusOneState, Optional[PerformanceProblem]]:
+        raise NotImplementedError
+
+    def finish(self) -> Optional[PerformanceProblem]:
+        return None
+
+    def _equivalent(self, a: Span, b: Span) -> bool:
+        """db spans are equivalent if their ops and hashes match. Other spans are
+        equivalent if their ops match."""
+        first_op = a.get("op") or None
+        second_op = b.get("op") or None
+        if not first_op or not second_op or first_op != second_op:
+            return False
+
+        if first_op.startswith("db"):
+            return a.get("hash") == b.get("hash")
+
+        return True
+
+
+class SearchingForMNPlusOne(MNPlusOneState):
+    """
+    The initial state for the MN+1 DB Query detector, and the state we return to
+    whenever there is no active repeating pattern being checked.
+
+    Keeps a list of recently seen spans until a repeat is found, at which point
+    it transitions to the ContinuingMNPlusOne state.
+    """
+
+    __slots__ = ("settings", "recent_spans")
+
+    def __init__(
+        self, settings: Dict[str, Any], initial_spans: Optional[Sequence[Span]] = None
+    ) -> None:
+        self.settings = settings
+        self.recent_spans = initial_spans.copy() if initial_spans else []
+
+    def next(self, span: Span) -> Tuple[MNPlusOneState, Optional[PerformanceProblem]]:
+        # Can't be a potential MN+1 without at least 2 previous spans.
+        if len(self.recent_spans) <= 1:
+            self.recent_spans.append(span)
+            return (self, None)
+
+        # Has an MN pattern begun to repeat itself? If so, transition to the
+        # ContinuingMNPlusOne state.
+        for i, recent_span in enumerate(self.recent_spans[:-1]):
+            if self._equivalent(span, recent_span):
+                pattern = self.recent_spans[i:]
```

Comment:
I'm trying to figure out why you need the `i` slicing here. Are there cases where the equivalent span is _not_ the first one in the recent spans? When that happens, doesn't slicing potentially omit spans from the pattern?

Reply:
So imagine a trace that goes `A B C D E C D E C D E`, where `D` is a DB query.

As the sixth span (index 5) span is checked, `recent_spans` will be `[A B C D E]`, and `span` will be `C`. When we enumerate over `recent_spans`, we'll find the equivalent span `C` at index 2. Therefore, the `pattern` we'll be looking to continue is `recent_spans[2:]`, or `[C D E]`. Does that make sense?

Reply:
Crystal clear, thank you!